src/Interpreter/Mulang.hs

```diff
@@ -0,0 +1,387 @@
+  | otherwise = do
+      v1 <- dereference r1
+      v2 <- dereference r2
+      createReference $ MuBool $ muEquals v1 v2
+
+muEquals (MuBool b1)   (MuBool b2)   = b1 == b2
+muEquals (MuNumber n1) (MuNumber n2) = n1 == n2
+muEquals (MuString s1) (MuString s2) = s1 == s2
+muEquals MuNull        MuNull        = True
+muEquals _             _             = False
+
+getParamNames :: [Mu.Pattern] -> [String]
+getParamNames params =
+  fmap (\(Mu.VariablePattern n) -> n) params
+
+runFunction :: [Reference] -> Mu.Expression -> Executable Reference
+runFunction functionEnv body = do
+  context <- get
+  returnValue <- callCC $ \(returnCallback) -> do
+    put (context { scopes = functionEnv, currentReturnCallback = returnCallback })
+    evalExpr body
+    return nullRef
+  modify' (\c -> c { scopes = scopes context
+                   , currentReturnCallback = currentReturnCallback context
+                   , currentRaiseCallback = currentRaiseCallback context
+                   , currentException = currentException context
+                   })
+  return returnValue
+
+findFrameForName :: String -> Executable Reference
+findFrameForName name = do
+  maybe (raiseString $ "Reference not found for name '" ++ name ++ "'") return
+    =<< findFrameForName' name
+
+findFrameForName' :: String -> Executable (Maybe Reference)
+findFrameForName' name = do
+  framesRefs <- gets scopes
+  frames :: [(Reference, Map String Reference)] <- forM framesRefs $ \ref -> do
+    dereference ref >>= \value -> case value of
+      (MuObject context) -> return (ref, context)
+      v -> error $ "Finding '" ++ name ++ "' the frame I got a non object " ++ show v
+
+  return $ fmap fst . find (Map.member name . snd) $ frames
+
+
+findReferenceForName :: String -> Executable Reference
+findReferenceForName name = do
+  ref <- findFrameForName name
+  (MuObject context) <- dereference ref
+  return $ context Map.! name
+
+getMaxKey :: Map k a -> Maybe k
+getMaxKey m = case Map.maxViewWithKey m of
+  Just ((k, _a), _) -> Just k
+  _ -> Nothing
+
+nullRef = Reference 0
+
+dereference' :: ObjectSpace -> Reference -> Value
+dereference' _ (Reference 0) = MuNull
+dereference' objectSpace ref = do
+  fromMaybe (error $ "Failed to find ref " ++ show ref ++ " in " ++ show objectSpace) .
+    Map.lookup ref $
+    objectSpace
+
+dereference :: Reference -> Executable Value
+dereference ref = do
+  objectSpace <- gets globalObjects
+  return $ dereference' objectSpace ref
+
+updateGlobalObjects f context =
+  context { globalObjects = f $ globalObjects context
+          }
+
+updateLocalVariables f context =
+  context { scopes = f $ scopes context
+          }
+
+incrementRef (Reference n) = Reference $ n + 1
+
+createReference :: Value -> Executable Reference
+createReference value = do
+  nextReferenceId :: Reference <- gets (fromMaybe (Reference 10) . fmap incrementRef . getMaxKey . globalObjects)
+  modify (updateGlobalObjects $ Map.insert nextReferenceId value)
```

Comment:
Why reference `10`?

Reply:
oh yeah, I saw that too. I don't think it has any special meaning, I think we could just replace that with a fromJust. I think globalObjects should never be empty.

Reply:
@ludat?

Reply:
Already talked with @julian-berbel and that 10 is useless, I think I replaced it with an exception but I never pushed that code. Initially objectSpace was empty but then I added contexts as objects and there is always at least one object alive so that 10 is never used